### PR TITLE
Service Fabric Managed Clusters - api version 2023-09-01-preview

### PR DESCRIPTION
**Pull request by a-santamaria**
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

## Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/azsdk/pr-suppressions) 
to get approval.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

**Comment by sjanamma** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ServiceFabric/managedClusters/{clusterName}/applications/{applicationName}/upgrade/read": {

This is not a valid path.

To fetch the application status you could possibly use this URI

"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ServiceFabric/managedClusters/{clusterName}/applications/{applicationName}/fetchUpgradeStatus"

---
Refers to: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/managedapplication.json:582 in 4b9057d. [](commit_id = 4b9057d82b663c73d3041375a92615af688db59f, deletion_comment = False)

**Comment by sjanamma** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ServiceFabric/managedClusters/{clusterName}/applications/{applicationName}/upgrade/read": {

/upgrade/ segment can be removed and action can be /rollbackUpgrade

    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ServiceFabric/managedClusters/{clusterName}/applications/{applicationName}/rollbackUpgrade: {


---
Refers to: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/managedapplication.json:582 in 4b9057d. [](commit_id = 4b9057d82b663c73d3041375a92615af688db59f, deletion_comment = False)

**Comment by sjanamma** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ServiceFabric/managedClusters/{clusterName}/applications/{applicationName}/upgrade/resume": {

Please review all the post call paths

---
Refers to: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/managedapplication.json:682 in 4b9057d. [](commit_id = 4b9057d82b663c73d3041375a92615af688db59f, deletion_comment = False)

**Comment by sjanamma** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ServiceFabric/managedclusters/{clusterName}/applications/{applicationName}/services/{serviceName}/partitions/readList": {

This is supposed to be List call. Also, the number of segments is more than what is allowed. You may need to model this differently to follow the standards.

---
Refers to: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/managedapplication.json:1085 in 4b9057d. [](commit_id = 4b9057d82b663c73d3041375a92615af688db59f, deletion_comment = False)

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/nodetype.json [hunk 1]
```diff
@@ -0,0 +1,1962 @@
...
+    "FrontendConfiguration": {
+      "type": "object",
+      "properties": {
+        "ipAddressType": {
+          "$ref": "#/definitions/IPAddressType",
+          "description": "The IP address type of this frontend configuration. If omitted the default value is IPv4."
+        },
+        "loadBalancerBackendAddressPoolId": {
+          "type": "string",
+          "description": "The resource Id of the Load Balancer backend address pool that the VM instances of the node type are associated with. The format of the resource Id is '/subscriptions/<subscriptionId>/resourceGroups/<resourceGroupName>/providers/Microsoft.Network/loadBalancers/<loadBalancerName>/backendAddressPools/<backendAddressPoolName>'."
```
Comment (sjanamma) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/nodetype.json:667]: format the property as arm-id 
[https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-arm-id-details](https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-arm-id-details)

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/nodetype.json [hunk 2]
```diff
@@ -0,0 +1,1962 @@
...
+        },
+        "loadBalancerInboundNatPoolId": {
+          "type": "string",
+          "description": "The resource Id of the Load Balancer inbound NAT pool that the VM instances of the node type are associated with. The format of the resource Id is '/subscriptions/<subscriptionId>/resourceGroups/<resourceGroupName>/providers/Microsoft.Network/loadBalancers/<loadBalancerName>/inboundNatPools/<inboundNatPoolName>'."
```
Comment (sjanamma) [inline specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/nodetype.json:671]: same format it as arm-id

**Comment by a-santamaria** [comment]
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ServiceFabric/managedClusters/{clusterName}/applications/{applicationName}/upgrade/read": {
> > ```
> 
> This is not a valid path.
> 
> To fetch the application status you could possibly use this URI
> 
> "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ServiceFabric/managedClusters/{clusterName}/applications/{applicationName}/fetchUpgradeStatus"
> 
> Refers to: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/managedapplication.json:582 in [4b9057d](https://github.com/Azure/azure-rest-api-specs/commit/4b9057d82b663c73d3041375a92615af688db59f). [](commit_id = [4b9057d](https://github.com/Azure/azure-rest-api-specs/commit/4b9057d82b663c73d3041375a92615af688db59f), deletion_comment = False)

Thanks for the review @sjanamma. Is it not valid because upgrade is not model as a resource?
We found this doc about actions https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#performing-an-action. Is there any other that we need to look at?

**Comment by a-santamaria** [comment]
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ServiceFabric/managedclusters/{clusterName}/applications/{applicationName}/services/{serviceName}/partitions/readList": {
> > ```
> 
> This is supposed to be List call. Also, the number of segments is more than what is allowed. You may need to model this differently to follow the standards.
> 
> lso, this is violating the multiple levels of nested types in this URI
> 
> Refers to: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/managedapplication.json:1085 in [4b9057d](https://github.com/Azure/azure-rest-api-specs/commit/4b9057d82b663c73d3041375a92615af688db59f). [](commit_id = [4b9057d](https://github.com/Azure/azure-rest-api-specs/commit/4b9057d82b663c73d3041375a92615af688db59f), deletion_comment = False)

@sjanamma what is the allowed number of segments? and what do you mean by "violating the multiple levels of nested types in this URI" ?

**Comment by a-santamaria** [comment]
@sjanamma I have reverted app actions. We will add them in the next api because we will need changes on our service

File: specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/nodetype.json (hunk 2 above)
Comment (a-santamaria) [reply specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/preview/2023-09-01-preview/nodetype.json:673]: fixed

**Comment by a-santamaria** [comment]
@rkmanda can you merge the pr please. its already signed off

**Review by raosuhas — APPROVED** [review]

**Comment by danielrbradley** [comment]
@raosuhas and @a-santamaria hello, I'm having some issues with downstream tooling having consumed this update 👋 

This version introduces two versions of the IP Tag definition - in `nodetype.json` and `managedcluster.json`.
In `managedcluster.json` it's named `#/definitions/IPTag`.
In `nodetype.json` it's named `#/definitions/IpTag`.

This causes an issue because both types exist in the same namespace, but only vary by case. We use these type names to generate files, but dependant on the case-insensitivity of the filesystem, these two files then conflict with one another.

Other versions of this specification only use the capitalized `IPTag` version. Would it be possible to upate the definition in `nodetype.json` to match?

**Pull request merged**